### ISSUE TITLE
BUG: Crash when attempting to search  

### ISSUE THREAD

**Issue by chrisdusovic** [Musish/Musish#493]
**Describe the bug**
When entering most search queries, the page crashes and displays the following error message:
"Oops, looks like something went wrong..."

**To Reproduce**

1. Enter"a" in the search bar.

**Expected behavior**
Results containing "a" appear.

**Screenshots**
Autocomplete results begin appearing, but the crash to the error page occurs before they are fully loaded.

**Desktop (please complete the following information):**

- OS: Windows 10 Pro 1909
- Browser: Chrome
- Version: 84.0.4147.105

**Additional context**
Here is the error in the console:
```
AlbumResultItem.tsx:38 Uncaught TypeError: Cannot read property 'name' of undefined
    at AlbumResultItem (AlbumResultItem.tsx:38)
    at renderWithHooks (react-dom.development.js:12938)
    at mountIndeterminateComponent (react-dom.development.js:15020)
    at beginWork (react-dom.development.js:15625)
    at performUnitOfWork (react-dom.development.js:19312)
    at workLoop (react-dom.development.js:19352)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at replayUnitOfWork (react-dom.development.js:18578)
```

I logged the results of the `library-albums` search and discovered that some of the results have identifiers but no attributes. These seem to be the culprit causing the crash.

```
href: "/v1/me/library/albums/l.xFync4A"
id: "l.xFync4A"
type: "library-albums"
```

When I try to query the Apple Music Library API for these albums I get a 404 error.

**Comment by napieralla** [comment]
Thanks for submitting such a detailed bug report.

Because you included the API endpoint that returned the results you crashed on, I could see that the error occurred when an album from your library was returned.
I was able to reproduce this by searching for album names in my library one character at a time and waiting for the results to load.
On some album titles I crash and on others I don't. Then on some I crash if I enter only one character but not if I enter two before giving the results time to load.

We could fix this either by performing checks whereever `MusicKit.MediaType` objects are used and ensuring `attributes` is truthy/object OR we could perform response validation on the MK API call before passing data further down the pipeline.
If we cannot do anything useful with the objects lacking `attributes` then I suggest the latter.

I tested this by querying some of the album `id`s that were returned to me without `attributes` through the `api.music.apple.com/v1/me/library/albums/` endpoint and received a `Resource with requested id was not found` error.
This suggests that it might be the APIs sending us bad data that is worth discarding.

For response validation we could use a library like [class-validator](https://github.com/typestack/class-validator) and run it against the `api.library.search` response [here](https://github.com/Musish/Musish/blob/01b9cbb1d8ddb0226fe407a915554021e1339344/src/app/components/Common/NavigationBar/Search/SearchBar.tsx#L121).

**Comment by raphaelvigee** [comment]
Thank you for that report.

It is a 'long' running known issue that has unfortunately not been resolved by any team member dues to a lack of time. 
Since we originally wrote Musish, it seems that Apple has slightly changed their API and some `attributes` dont exist (we never encountered that issue during the active development of Musish).

Here are some of the reports that have been collected from related issues:

https://sentry.io/share/issue/c92b43ee82de4790ad7c0164b4fa3ff4/
https://sentry.io/share/issue/d550faf8d5f049999536fd3448c3f431/
https://sentry.io/share/issue/8cd67953e5d54610a3946cb30baa4324/
https://sentry.io/share/issue/5dfcd153365f45449650c9afa40177d3/

We would happily accept PRs resolving such issues.

**Comment by raphaelvigee** [comment]
> For response validation we could use a library like class-validator and run it against the api.library.search response here.

I think a better approach would be to wrap any access to `xxx.attributes` with some utils function that allows for a fallback. This solution is lightweight, controlled and prevent Musish to become (even more) of a dependencies hell.

**Comment by napieralla** [comment]
Thanks for the Sentry reports, @raphaelvigee.

> I think a better approach would be to wrap any access to xxx.attributes with some utils function that allows for a fallback. This solution is lightweight, controlled and prevent Musish to become (even more) of a dependencies hell.

Are you thinking of something like this?
````
function validateMediaItem(item) {
  if(!item.attributes) return { ...item, attributes: { name: 'Unkown item' } };
  else return item;
}
````

The issues I see with this approach are:

1. It doesn't scale well; we have to do these checks whereever media-items returned from the APIs are used.
2. Risk of negligence; we or future contributors might forget to do these checks somewhere.
3. Unnecessary output; if we don't know what is causing some media-items to return without `attributes` then showing a 'fallback'-item seems pointless and might confuse users.

I think validating fetched data as soon as it arrives from the API is important.
This would allow us to rely on the compile-time type-security that Typescript provides.

I hope to hear your thoughts on this.

**Comment by raphaelvigee** [comment]
I was thinking more of something like that:
```
function itemAttr(item: object, attr: string, fallback: any) {
    if (attr in item && item[attr]) {
        return item[attr]
    } else {
        return fallback
    }
}
```

But now there is an issue that needs addressing: if the attributes are missing, how do we display the items? Like, is there an other object property available to us to display the items in a human manner (and not have loads of songs called `N/A`)?
- If yes, then the `itemAttr` function should be responsible for the magic
- If no, maybe we should filter out the results without attributes (aka #494) 

Regarding your concerns, there is a risk of negligence, CR should prevent that, maybe having some rule as part of the CI (no direct access to `Item.attributes`)

**Comment by napieralla** [comment]
I had not noticed that `MusicKit.MediaItem.attributes` was defined as `any` by the `musickit-typescript` package. I am used to working with stricter (ESLint/tsconfig) rules that would never have allowed `attributes.something` to be accessed without a runtime check in the first place.

After this late realization I think it would be better **not** to filter `MediaItem` results that lack `attributes`, but to rather allow the individual components to decide on how to deal with this. That is line with what you (@raphaelvigee) were suggesting with the `itemAttr` function.

In conjunction with this we need to disallow the use of `any`.
This can be done with ESLint (handles TS now too, TSLint is deprecated) and maybe with stricter `tsconfig.json` rules.
That involves quite a bit of work but it would allow us to catch these kind of errors before they arrive as Sentry logs.

**Comment by raphaelvigee** [comment]
> I had not noticed that MusicKit.MediaItem.attributes was defined as any by the musickit-typescript package. I am used to working with stricter (ESLint/tsconfig) rules that would never have allowed attributes.something to be accessed without a runtime check in the first place.

Yeah exactly, that's part of the problem, Apple doesnt bless us with some typescript, so some dude made that, its useful, but not perfect.

> After this late realization I think it would be better not to filter MediaItem results that lack attributes, but to rather allow the individual components to decide on how to deal with this. That is line with what you (@raphaelvigee) were suggesting with the itemAttr function.

But the issue is, if we dont have `attributes`, how tf do we display names etc, filtering out may be our only solution

> In conjunction with this we need to disallow the use of any.
This has been the plan since the introduction of TS, but again... lack of time

We also need to move away from TSlint, but thats another story. The current rules are kinda loose because the conversion to TS was done kindof in a rush, stricter rules would have made stuff really hard, especially since we dont have strict typing for all MusicKit stuff.

**Comment by chrisdusovic** [comment]
A potential solution to this is to wrap the MusicKit API in an abstraction layer that validates the response objects and filters out the bad ones. The objects could then be mapped to TypeScript interfaces as well.

**Comment by napieralla** [comment]
> We also need to move away from TSlint, but thats another story. The current rules are kinda loose because the conversion to TS was done kindof in a rush, stricter rules would have made stuff really hard, especially since we dont have strict typing for all MusicKit stuff.

I will look into this later and see if I can provide a comprehensive fix for these type-related bugs, by adding stricter checks and updating all modules to adhere to them.

> A potential solution to this is to wrap the MusicKit API in an abstraction layer that validates the response objects and filters out the bad ones. The objects could then be mapped to TypeScript interfaces as well.

That would be a very elegant way of interfacing with MusicKit, but MK is too deeply intertwined with the functionality of different components for it to be manageable without a serious re-write.

**Comment by chrisdusovic** [comment]
The abstraction layer could completely mirror the MusicKit API so components would be completely unaware of the change. Validation and TypeScript conversion could happen internally.

**Comment by napieralla** [comment]
> The abstraction layer could completely mirror the MusicKit API so components would be completely unaware of the change. Validation and TypeScript conversion could happen internally.

Is this better than what we have at the moment? We can still validate data and safely convert to Typescript with our current system, while being tightly integrated with MusicKit's APIs.

The benefits that I see in further abstraction are allowing us to integrate with other music-streaming services and improving code clarity. But I don't see how this would help us solve issues like the current one.
Could you elaborate on your thoughts on this and the benefits you see?

**Comment by chrisdusovic** [comment]
It seems to me that in the current system, validation must be done by each individual component. I am assuming that validation will happen similar to my hotfix (#494). The way I was imagining it was to create a wrapper around MusicKit that does validation and TypeScript conversion and inject it [here](https://github.com/Musish/Musish/blob/8eb23031899ced29cce51e6e2dd0321819998974/src/app/hoc/withMK.tsx#L47). Components will get the new benefits without any changes required.

**Comment by raphaelvigee** [comment]
Maybe injecting the wrapper in `withMK` is a good idea, but if we narrow down the types, changes will be required down in the components.

Keep in mind there is also [useMK](https://github.com/Musish/Musish/blob/8eb23031899ced29cce51e6e2dd0321819998974/src/app/hooks/useMK.tsx)
